File: 1_semestr/6/Stack/Stack/stack.cpp
#include <stdio.h>
#include <iostream>
#include "stack.h"

struct ElementOfStack
{
	ValueType value;
	ElementOfStack *next;
};

struct Stack
{
	ElementOfStack *head;
	SizeType size;
};

Stack *createNewStack()
{
	Stack *result = new Stack;
	result->size = 0;
	result->head = nullptr;
	return result;
};

SizeType size(Stack *stack)
{
	return stack->size;
}

ValueType top(Stack *stack, bool &test)
{
	if (stack->size != 0)
	{
		return stack->head->value;
	}
	else
	{
		printf("Error: stack is empty\n");
		test = true;
		return -99999;
	}
}

void push(Stack *stack, const ValueType value)
{
	ElementOfStack *newElement = new ElementOfStack;
	newElement->value = value;
	newElement->next = stack->head;
	stack->head = newElement;
	stack->size++;
}

void pop(Stack *stack, bool &test)
{
	if (stack->size != 0)
	{
		Position temp = stack->head;
		stack->head = stack->head->next;
		stack->size--;
		delete temp;
	}
	else
	{
		printf("Stack is empty\n");
		test = 1;
	}
}

void deleteStack(Stack *stack)
{
	bool test = false;
	while (stack->head != nullptr)
	{
		pop(stack, test);
	}
	if (test)
	{
		printf("Error");
	}
	delete stack;
}

File: 2_semestr/7/1/Stack/Stack.cs
using System;

namespace Stack
{
    public class Stack<T>
    {
        private ElementOfStack head;

        /// <summary>
        /// Size of stack.
        /// </summary>
        public int Size { get; set; }

        /// <summary>
        /// Stack's container.
        /// </summary>
        private class ElementOfStack
        {
            public T Value { get; set; }
            public ElementOfStack Next { get; set; }
            public ElementOfStack(Stack<T> stack)
            {
                this.Next = stack.head;
            }
        }

        public Stack()
        {
            head = null;
            Size = 0;
        }

        /// <summary>
        /// Push element to top.
        /// </summary>
        /// <param name="value"></param>
        public void Push(T value)
        {
            ElementOfStack newElement = new ElementOfStack(this);
            newElement.Value = value;
            head = newElement;
            Size++;
        }

        /// <summary>
        /// Delete element from top.
        /// </summary>
        public void Pop()
        {
            if (head != null)
            {
                this.head = this.head.Next;
                Size--;
            }
            else
            {
                throw new EmptyStackException();
            }
        }

        /// <summary>
        /// Write element from top.
        /// </summary>
        /// <returns></returns>
        public T Top()
        {
            if (head != null)
            {
                return head.Value;
            }
            else
            {
                throw new EmptyStackException();
            }
        }

        /// <summary>
        /// Clear stack.
        /// </summary>
        public void Clear()
        {
            head = null;
            Size = 0;
        }
    }
}

